src/coreclr/gc/gc.cpp

```diff
@@ -10939,44 +10952,22 @@ heap_segment* gc_heap::get_free_region (int gen_number, size_t size)
     if (gen_number <= max_generation)
     {
         assert (size == 0);
-        if (free_regions)
-        {
-            num_free_regions--;
-            num_free_regions_removed++;
-            region = free_regions;
-            dprintf (REGIONS_LOG, ("%d free regions left, get %Ix-%Ix-%Ix",
-                num_free_regions, heap_segment_mem (region),
-                heap_segment_committed (region), heap_segment_used (region)));
-            free_regions = heap_segment_next (free_regions);
-            committed_in_free -= heap_segment_committed (region) - get_region_start (region);
-        }
+        region = free_regions[basic_free_region].unlink_region_front();
     }
     else
     {
-        // look for a region that is large enough
-        heap_segment** link = &free_large_regions;
-        for (region = free_large_regions; region != nullptr; region = heap_segment_next (region))
-        {
-            uint8_t* region_start = get_region_start(region);
-            uint8_t* region_end = heap_segment_reserved(region);
+        const size_t LARGE_REGION_SIZE = global_region_allocator.get_large_region_alignment();
 
-            if ((region_end - region_start) >= (ptrdiff_t)size)
-            {
-                // found a region that is large enough
-                break;
-            }
-            link = &heap_segment_next (region);
-        }
+        free_region_kind kind = size <= LARGE_REGION_SIZE ? large_free_region : huge_free_region;
+
+        // get it from the local list of free large regions if possible
+        region = free_regions[kind].unlink_region_front();
```

Comment:
if it's a huge region the first region on the huge region list may not be large enough

Reply:
but then again, would there be any regions that are actually on the per heap huge region list, since we move them all to global in every GC?

Reply:
but then again, would there by anything on the perheap huge region list at all since we move them all to the global list in every GC? it seems like the perheap list is only temporarily there between `thread_final_regions` and `distribute_free_regions".

Reply:
You're right on both counts. I think I should slightly restructure the code so it's safe even if the huge region free list per heap is non-empty.